### PR TITLE
Allow non-formated seconds to be included as a value of the itunes:duration element

### PR DESCRIPTION
**Pull request by aitor**
As mentioned in ruby/rss#4 the iTunes Podcast Spec supports the use of nonformated seconds as the value for `itunes:duration` elements:

> If you specify a single number as a value (without colons), Apple Podcasts displays the value as seconds.

source: [RSS tags for Podcasts Connect](https://help.apple.com/itc/podcasts_connect/#/itcb54353390)

I've updated the regex used in `ITunesDuration#parse` to accept any string of digits and reflect the spec point in the code by calculating from the seconds the corresponding components if no colons are present.

I've followed the "be conservative in what you send, be liberal in what you accept" [principle](https://www.wikiwand.com/en/Robustness_principle) and if a maker creates a feed by speciying just seconds:

``` rb
      item.itunes_duration = "315"
```

It will still output formatted values (eg. `05:15`).

### CODE REVIEW

File: test/rss/test_itunes.rb [hunk 1]
```diff
@@ -203,8 +203,9 @@ def assert_itunes_duration(readers, &rss20_maker)
         _assert_itunes_duration(7, 14, 5, "7:14:05", readers, &rss20_maker)
         _assert_itunes_duration(0, 4, 55, "04:55", readers, &rss20_maker)
         _assert_itunes_duration(0, 4, 5, "4:05", readers, &rss20_maker)
+        _assert_itunes_duration(0, 0, 5, "5", readers, &rss20_maker)
+        _assert_itunes_duration(0, 3, 15, "195", readers, &rss20_maker)
 
-        _assert_itunes_duration_not_available_value("5", &rss20_maker)
```
Comment (kou) [inline test/rss/test_itunes.rb]: Could you add  a test for an empty string case at the last of this test method?

File: test/rss/test_maker_itunes.rb [hunk 2]
```diff
@@ -242,21 +242,17 @@ def _assert_maker_itunes_duration_invalid_value(value, maker_readers)
...
-                                      feed_readers)
-        _assert_maker_itunes_duration(0, 4, 5, "4:05", maker_readers,
-                                      feed_readers)
-
-        _assert_maker_itunes_duration_invalid_value("5", maker_readers)
+        _assert_maker_itunes_duration(7, 14, 5, "07:14:05", maker_readers, feed_readers)
+        _assert_maker_itunes_duration(7, 14, 5, "7:14:05", maker_readers, feed_readers)
+        _assert_maker_itunes_duration(0, 4, 55, "04:55", maker_readers, feed_readers)
+        _assert_maker_itunes_duration(0, 4, 5, "4:05", maker_readers, feed_readers)
+        _assert_maker_itunes_duration(0, 0, 5, "0:05", maker_readers, feed_readers)
```
Comment (kou) [inline test/rss/test_maker_itunes.rb]: Could you remove style changes? It's not the main subject of this pull request.
These changes over 80 columns in one line.

File: test/rss/test_maker_itunes.rb [hunk 3]
```diff
@@ -242,21 +242,17 @@ def _assert_maker_itunes_duration_invalid_value(value, maker_readers)
...
+        _assert_maker_itunes_duration_by_value(0, 5, 15, "315", maker_readers, feed_readers)
+
         _assert_maker_itunes_duration_invalid_value("09:07:14:05", maker_readers)
         _assert_maker_itunes_duration_invalid_value("10:5", maker_readers)
         _assert_maker_itunes_duration_invalid_value("10:03:5", maker_readers)
         _assert_maker_itunes_duration_invalid_value("10:3:05", maker_readers)
-
```
Comment (kou) [inline test/rss/test_maker_itunes.rb]: Could you remove style changes?
This empty line exists for separating invalid value classes.
"10:3:05" is an invalid value for "valid characters but invalid format" class.
"xx:xx:xx" is an invalid value for "valid format but invalid characters" class.

File: test/rss/test_maker_itunes.rb [hunk 4]
```diff
@@ -242,21 +242,17 @@ def _assert_maker_itunes_duration_invalid_value(value, maker_readers)
...
-        _assert_maker_itunes_duration(0, 4, 5, "4:05", maker_readers,
-                                      feed_readers)
-
-        _assert_maker_itunes_duration_invalid_value("5", maker_readers)
+        _assert_maker_itunes_duration(7, 14, 5, "07:14:05", maker_readers, feed_readers)
+        _assert_maker_itunes_duration(7, 14, 5, "7:14:05", maker_readers, feed_readers)
+        _assert_maker_itunes_duration(0, 4, 55, "04:55", maker_readers, feed_readers)
+        _assert_maker_itunes_duration(0, 4, 5, "4:05", maker_readers, feed_readers)
+        _assert_maker_itunes_duration(0, 0, 5, "0:05", maker_readers, feed_readers)
+        _assert_maker_itunes_duration_by_value(0, 5, 15, "315", maker_readers, feed_readers)
```
Comment (kou) [inline test/rss/test_maker_itunes.rb]: Could you also add a test for "3600" over case?

File: lib/rss/itunes.rb [hunk 5]
```diff
@@ -277,33 +277,46 @@ def required_uri
         def parse(duration, do_validate=true)
           if do_validate and /\A(?:
                                   \d?\d:[0-5]\d:[0-5]\d|
-                                  [0-5]?\d:[0-5]\d
+                                  [0-5]?\d:[0-5]\d|
+                                  \d*
```
[the remark this answers is not in the document]
Comment (aitor) [reply lib/rss/itunes.rb]: Right.

File: lib/rss/itunes.rb [hunk 6]
```diff
@@ -277,33 +277,46 @@ def required_uri
...
                                 )\z/x !~ duration
             raise ArgumentError,
-                    "must be one of HH:MM:SS, H:MM:SS, MM::SS, M:SS: " +
+                    "must be one of (HH:MM:SS, H:MM:SS, MM:SS, M:SS, SS) : " +
```
[the remark this answers is not in the document]
Comment (aitor) [reply lib/rss/itunes.rb]: > Seconds only case doesn't restrict 2 digits. So S+ or something is better.

I thougt that given it's just a error info, an general indication suggesting seconds are ok was enough… without going regex-level crazy. I've updated it to`S+`.

> BTW, why did you add "(" ... ")"? I think that "HH:MM:SS, ... or S+: " is better.

Because `:` is used to separate the components of the duration too, so the last one used for gramatical purposes, I think THAT was pretty confusing. I applied the parenthesis to provide clear separation… I'll remove it.

File: test/rss/test_maker_itunes.rb (hunk 4 above)
Comment (aitor) [reply test/rss/test_maker_itunes.rb]: Ok.

File: test/rss/test_maker_itunes.rb (hunk 3 above)
Comment (aitor) [reply test/rss/test_maker_itunes.rb]: Ok.

File: test/rss/test_itunes.rb (hunk 1 above)
Comment (aitor) [reply test/rss/test_itunes.rb]: Ok.

File: lib/rss/itunes.rb [hunk 7]
```diff
@@ -277,33 +277,46 @@ def required_uri
...
+          return unless minutes && seconds
+
+          components = seconds_to_components(seconds)
+          components[0] += hours
+          components[1] += minutes
+          components.shift unless components[0] > 0
+          format_duration(components)
+        end
+
+        def seconds_to_components(total_seconds)
```
[the remark this answers is not in the document]
Comment (aitor) [reply lib/rss/itunes.rb]: Ok.

File: test/rss/test_maker_itunes.rb (hunk 2 above)
Comment (aitor) [reply test/rss/test_maker_itunes.rb]: It would be great if you could provide a rubocop file to make the styles rules explicit. 80 cols seem archaic to me in 2019 but a linter will take care of this automatically and set the expectations right. I've removed the changes.

File: test/rss/test_maker_itunes.rb (hunk 3 above)
Comment (aitor) [reply test/rss/test_maker_itunes.rb]: PS: I'd say that different invalid value classes should have different tests to check them. Using an empty line to separate them is too subtle at best.

File: test/rss/test_itunes.rb [hunk 8]
```diff
@@ -203,8 +203,9 @@ def assert_itunes_duration(readers, &rss20_maker)
         _assert_itunes_duration(7, 14, 5, "7:14:05", readers, &rss20_maker)
         _assert_itunes_duration(0, 4, 55, "04:55", readers, &rss20_maker)
         _assert_itunes_duration(0, 4, 5, "4:05", readers, &rss20_maker)
+        _assert_itunes_duration(0, 0, 5, "5", readers, &rss20_maker)
+        _assert_itunes_duration(0, 3, 15, "195", readers, &rss20_maker)
```
[the remark this answers is not in the document]
Comment (aitor) [reply test/rss/test_itunes.rb]: Ok.

**Comment by aitor** [comment]
@kou check the PR… I think I've updated all the requested changes.

File: lib/rss/itunes.rb [hunk 9]
```diff
@@ -277,34 +277,51 @@ def required_uri
...
+        def construct(hours, minutes, seconds)
+          components = [minutes, seconds]
           if components.include?(nil)
             nil
           else
-            components.unshift(hour) if hour and hour > 0
-            components.collect do |component|
-              "%02d" % component
-            end.join(":")
+            components = seconds_to_components(seconds)
```
Comment (kou) [inline lib/rss/itunes.rb]: Do we need "60" over seconds support here?
This method accepts hours, minutes and seconds instead of one "duration value" in string. So I think that we can assume that seconds is less than 60.

It's OK to me that `RSS::Maker::ITunesItemModel;:ITunesDurationBase#second=` accepts "60" over seconds.

File: lib/rss/itunes.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (kou) [reply lib/rss/itunes.rb]: > I applied the parenthesis to provide clear separation

OK. It makes sense.

File: test/rss/test_maker_itunes.rb (hunk 3 above)
Comment (kou) [reply test/rss/test_maker_itunes.rb]: > different invalid value classes should have different tests to check them.

I think so too.
But it's out of scope of this pull request.

File: lib/rss/itunes.rb (hunk 9 above)
Comment (aitor) [reply lib/rss/itunes.rb]: I'm not sure what you mean here. `RSS::ITunesItemModel::ITunesDuration.construct` will construct a correct string for any value of seconds under or over 60, mixed with other components or not.

```
RSS::ITunesItemModel::ITunesDuration.construct(0,0,61)
=> "01:01"
RSS::ITunesItemModel::ITunesDuration.construct(0,0,3661)
=> "01:01:01"
RSS::ITunesItemModel::ITunesDuration.construct(0,1,61)
=> "02:01"
RSS::ITunesItemModel::ITunesDuration.construct(0,1,3661)
=> "01:03:01"
RSS::ITunesItemModel::ITunesDuration.construct(1,1,61)
=> "01:02:01"
RSS::ITunesItemModel::ITunesDuration.construct(1,1,3661)
=> "02:02:01"
```

Comment (kou) [reply lib/rss/itunes.rb]: I wanted to say `RSS::ITunesItemModel::ITunesDuration.construct(0,0,61)` should be an invalid argument case. Because "00:00:61" is an invalid value.

Comment (kou) [reply lib/rss/itunes.rb]: Do you have any real world use case of this method?
It's designed to be used only from `RSS::Maker::ITunesItemModel;:ITunesDurationBase#update_content`. We can assume that users specify all hour, minute and second there. Users never use `RSS::Maker::ITunesItemModel;:ITunesDurationBase#update_content` directly.

Comment (aitor) [reply lib/rss/itunes.rb]: Well, it’s exposed publicly so people can/will use it. To directly answer your question I don’t have a real world example of using this specific method. Even then I think makes sense for any method to give the best possible answer for any given input. 

Anyway, what do you want me to do then in this method?

Comment (kou) [reply lib/rss/itunes.rb]: I want you to remove this change.

I don't want provide too much features without any reasonable use case. It increases maintenance cost. I need to keep maintaining this library.

Comment (aitor) [reply lib/rss/itunes.rb]: I've removed the support for 60+ seconds in `RSS::ITunesItemModel::ITunesDuration#construct` Let me know if you like it now.
